### DISCUSSION TITLE
Is there some symbolic link support by API level in sbt?

### DISCUSSION THREAD

**Discussion by djx314** [sbt/sbt#7024] [Q&A] [answered]
https://github.com/ThoughtWorksInc/dsl-domains-cats/blob/c24d1aeeea8748b5dffe2988321fb930b89433c3/domains-cats/.jvm/build.sbt#L1

with a symbolic link in linux file system, we can easy to defined scala.js and scala.jvm depts in a simple way.  
Is there some support in sbt so that we can use it in windows and any sbt supported system in the same way?

**Comment by eed3si9n** [comment]
You mean do we emulate symbolic link traversal on Windows? I don't think so.

**Reply by djx314** [reply]
emmm, not exactly. You can understand it as one `build.sbt` can copy another `build.sbt` file as a string that import it to the specific place. Just like
```scala
// build.sbt
includeSbtFileAsCopyScriptString("../../build.sbt")
```
Or something like javascript's `eval(scriptString)`.

**Reply by djx314** [reply]
I thought about this problem again. It seems that it just emphasizes `build.sbt` is a script. And it's easy to know if we can use `includeAsString`, many setting in project/xx.scala now calling as a function return by `Seq[Setting[_]]` can just implementing like `includeAsString("xx/build.sbt.script")`.

**Reply by eed3si9n** [reply]
sbt's AutoPlugin is basically an abstraction around `Seq[Setting[_]]`, which you can locally define in `project/*.scala`.
For JS vs JVM, I like using https://github.com/sbt/sbt-projectmatrix.

**Comment by djx314** [comment]
@eed3si9n Finally, this requirement may be such a simple function that sbt itself may provide both. It's just that I couldn't reasonably transform this problem for a while.  
If a sbt project can specify one or more additional, specified build files when project creating, that is ok.
```scala
val pro = (project in file(".") / "simple-adt").addBuildScript(file(".") / ".." / "build.sbt").addBuildScript(file(".") / ".." / "depts.sbt")
```

**Reply by eed3si9n** [reply]
There used to be a feature like that called settingSet, but we dropped it in sbt 1.x, I guess because it's kind of like AutoPlugins - https://github.com/sbt/sbt/pull/3068

**Reply by djx314** [reply]
It seems that `BasicKeys.extraMetaSbtFiles` can be use in this problem.  
Is there any mthod to chang this `AttributeKey` in AutoPlugin that I can set the value before a project load and I can no need to dependency a low level API in `build.sbt`?

**Reply by djx314** [reply]
Or any AutoPlugins can write as something like `build.sbt` can also solve my problem.  
Just some configure like `libraryDependencies +=` need to be in `*.sbt` and these configuration need to be use in multiple places.  
I have tried `Thisbuild` with scope, but not friendly.

**Comment by djx314** [comment]
```scala
// CopyFilePlugin .scala
trait WithFile {
  self =>

  protected def baseFile: File

  protected def extraFiles: Seq[File]

  def write: Unit = write(baseFile)

  def write(f1: File, f: File*): Unit = write(f1 +: f)

  def write(f: Seq[File]): Unit = {
    val arrs = extraFiles.map(t => Files.readAllBytes(t.toPath))
    for (extFile <- f) {
      val copyFile = extFile / "copyBuild.sbt"
      Files.deleteIfExists(copyFile.toPath)
      Using.resource(new java.io.FileOutputStream(copyFile)) { p =>
        Using.resource(new PrintWriter(p)) { u =>
          for (eachArr <- arrs) {
            p.write(eachArr)
            p.flush()
            u.println()
            u.flush()
          }
        }
      }
    }
  }

  def write(f1: File => File, f: (File => File)*): Unit = write((f1 +: f).map(t => t(self.baseFile)))

  def withBuildFile(f: Seq[File]): WithFile = new WithFile {
    override def baseFile: File        = self.baseFile
    override def extraFiles: Seq[File] = f ++: self.extraFiles
  }

  def withBuildFile(f1: File, f: File*): WithFile = withBuildFile(f1 +: f)

  def withBuildFile(f1: File => File, f: (File => File)*): WithFile = withBuildFile((f1 +: f).map(t => t(self.baseFile)))
}

object WithFile {
  def apply(f: File): WithFile = new WithFile {
    override def baseFile: File        = f
    override def extraFiles: Seq[File] = Seq.empty
  }
}

implicit class fileExt(val file: File) extends AnyVal {
  def withBuildFile(f: Seq[File]): WithFile                         = WithFile(file).withBuildFile(f)
  def withBuildFile(f1: File, f: File*): WithFile                   = WithFile(file).withBuildFile(f1, f: _*)
  def withBuildFile(f1: File => File, f: (File => File)*): WithFile = WithFile(file).withBuildFile(f1, f: _*)
}
```
```scala
// build.sbt
val toJsRoot: File => File  = f => f / "js"
val toJVMRoot: File => File = f => f / "jvm"

val `root-path`    = file(".").getAbsoluteFile
val `scripts-root` = `root-path` / "scripts"
val globalScripts  = Seq(`scripts-root` / "all-depts.sbt.script", `scripts-root` / "settings-global.sbt.script")

val scalaJsCommonScripts = Seq(`scripts-root` / "scalajs-common.sbt.script")
val scalaJsJvmScripts    = scalaJsCommonScripts
val scalaJsJsScripts     = Seq(`scripts-root` / "scalajs-js.sbt.script") ++: scalaJsCommonScripts

val initCopyFile = {
  `adt-path`.withBuildFile(globalScripts).write
  `adt-codegen-path`.withBuildFile(globalScripts).write
  `injection-path`.withBuildFile(globalScripts).write
  locally {
    val filePath = `adt-core-path`.withBuildFile(globalScripts).withBuildFile(scalaJsJvmScripts).withBuildFile(_ / "build.sbt.script")
    filePath.write(toJVMRoot)
    filePath.withBuildFile(scalaJsJsScripts).write(toJsRoot)
  }
  locally {
    val filePath = `injection-core-path`.withBuildFile(globalScripts).withBuildFile(scalaJsJvmScripts).withBuildFile(_ / "build.sbt.script")
    filePath.write(toJVMRoot)
    filePath.withBuildFile(scalaJsJsScripts).write(toJsRoot)
  }
  locally {
    val filePath = `test-common-path`.withBuildFile(globalScripts).withBuildFile(scalaJsJvmScripts).withBuildFile(_ / "build.sbt.script")
    filePath.write(toJVMRoot)
    filePath.withBuildFile(scalaJsJsScripts).write(toJsRoot)
  }
}
```
@eed3si9n Finally solve it in a simple way. 🤣

**Reply by djx314** [reply]
Just copy string to copyBuild.sbt before project load.

**Comment by djx314** [comment]
For days passed. It works well for my project. Just pay attention to add these copied files to `.gitignore` and `.scalafmt.conf` -> `project.excludePaths`

**Comment by djx314** [comment, marked as answer]
https://github.com/ThoughtWorksInc/dsl-domains-cats/blob/master/project/plugins.sbt.scalajs06#L8-L22

Find a more simple way.  
1. Write a build.sbt
1. Use the template(String concat) to gen it to a AutoPlugin source file.
1. Add the source file to the project.
1. Just enable the plugin(enalbe the settings) when it's needed.

**Reply by djx314** [reply]
The [implementation link](https://github.com/sbt/sbt/discussions/7024#discussioncomment-3790761).

**Comment by djx314** [comment]
This is a reference implementation of [the describe](./7024#discussioncomment-3775846) above.

↓ A windows compatible implementation

https://github.com/scalax/simple-nat/blob/36bafb5d0d612e23006e28f28bac74dc57731876/modules/main/simple-adt/core/js/add-plugin.sbt
https://github.com/scalax/simple-nat/tree/36bafb5d0d612e23006e28f28bac74dc57731876/scripts
https://github.com/scalax/simple-nat/blob/36bafb5d0d612e23006e28f28bac74dc57731876/project/plugin.sbt#L19-L57

↑ A windows compatible implementation  
↓ based on

https://github.com/ThoughtWorksInc/dsl-domains-cats/blob/f57535968e03944853920fe63dca661b5e65b182/domains-cats/.js/build.sbt
https://github.com/ThoughtWorksInc/dsl-domains-cats/blob/f57535968e03944853920fe63dca661b5e65b182/project/plugins.sbt.scalajs06#L8-L22

↑ based on
